### PR TITLE
App strings not being translated for Indonesian and Hebrew

### LINKED ISSUE #18029: App language not working for Indonesian
This issue is only reproducible on JP and it seems like it started on version `20.6`.
It works as expected on the previous version of the Locale picker (tested versions `17.5`, `19.4` and `19.8.1`) but it was also working for some time after the new Locale picker was created (tested versions `20.2` and `20.5`).

### Expected behavior
When Indonesian is selected on "App Settings", the app strings should be translated to Indonesian.

### Actual behavior
When Indonesian is selected on "App Settings", some of the strings in the Locale picker are displayed in Indonesian, as expected, but most of the app strings are displayed in English.

### Steps to reproduce the behavior
1 - Install JP starting from version `20.6`;
2 - Tap the user picture to open "Me" screen;
3 - Tap on "App Settings";
4 - Tap on "Interface Language";
5 - Select Indonesian.

##### Tested on emulator, Android 12, JPAndroid `21.7`.

### PR DESCRIPTION
**Pull request by RenanLukas**
Fixes #18029 

To test:
1 - Install JP app;
2 - Tap the user picture to open "Me" screen;
3 - Tap on "App Settings";
4 - Tap on "Interface Language";
5 - Select Indonesian and navigate through the app: the strings should be translated to Indonesian;
6 - Repeat steps 2 - 4 and select Hebrew language. Navigate through the app: the strings should be translated to Hebrew.

Thanks @ParaskP7 for proposing the code changes on this fix.

<p float="left">
  
   
</p>

## Regression Notes
1. Potential unintended areas of impact
None

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
`-`

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

**Review by ravishanker — APPROVED** [review]
Thanks for taking care of this🚀 👍 

See, if this can go into 22.4

**Review by ParaskP7 — CHANGES_REQUESTED** [review]
👋 @RenanLukas and thanks for taking care of this! 🥇 

I have reviewed and tested this, everything (almost) LGTM! 💯 

Let me elaborate on the (almost) part and also ask guidance from @AliSoftware as I am failing to make the connection between `values-he/strings.xml` -> `values-iw/strings.xml` (for Hebrew) and `values-id/strings.xml` -> `values-in/strings.xml` (for Indonesia). 😊 

1. You added the empty `src/jetpack/res/values-iw/strings.xml` and `src/jetpack/res/values-in/strings.xml`, but I think you should instead copy-paste the `src/jetpack/res/values-he/strings.xml` and `src/jetpack/res/values-id/strings.xml` content respectively. To my understanding the `values-iw/strings.xml` should be an identical copy of `values-he/strings.xml`, same applies to `values-in/strings.xml`, which should be an identical copy of `values-id/strings.xml`. Correct me if I am wrong @AliSoftware . 🙏 
2. Now, even if you do the above, we need to make sure that the automatic GlotPress related  process is keeping them in sync for Jetpack, just as it happen for WordPress already. I quickly tried to figure out where that logic exist, but couldn't. @AliSoftware can you help guide us here? 🙏 
3. As per @AliSoftware (Slack) comment, we have to keep both `he` + `iw`, as well as `id` + `in`, but I am not sure how this is done, is it a symlink of some kind, is it something else. No matter, this process needs to work for Jetpack, just like it currently works on WordPress, as least for the time being and until we do another round of checks to figure out whether those old `he` and `id` locale code are still in use (depending on the version of Java/Android/SDK, or used at runtime/compile time), and if not, to completely remove them if that is not relevant anymore, which will avoid all the above dance.

Please let me know your thoughts and if the above make sense. In the meanwhile, I'll request changes just to refrain from committing those empty `strings/xml` files as I am not sure how this will effect the overall localization solution production-wise. 🤔

**Comment by AliSoftware** [comment]
Yes in the main flavor those are symlinks, that's why we don't even need to have any logic implemented to copy them every time they change or to keep them in sync, that's automatically provided by the magic of symlinks 🙂
 
You can run `ls -lF sec/main/res` to see how it's done in the main flavor and then replicate the same type of symlinks in `src/Jetpack/res` from there.

**Comment by ParaskP7** [comment]
Thanks for the reply @AliSoftware !

> Yes in the main flavor those are symlinks, that's why we don't even need to have any logic implemented to copy them every time they change or to keep them in sync, that's automatically provided by the magic of symlinks 🙂

Magic! 🪄 🪄 🪄

So, to my understanding, my suggestion is NOT that correct them, we should not create by copy-pasting the content of those 2 Jetpack `he` and `id` localization files into their related `values-iw/strings.xml` and `values-in/strings.xml` newly created (by @RenanLukas) files, but instead delete those files and create a symlink instead, right? 🤔 

> You can run ls -lF sec/main/res to see how it's done in the main flavor and then replicate the same type of symlinks in src/Jetpack/res from there.

@RenanLukas can you take care of that? 😊

**Comment by RenanLukas** [comment]
Thanks @ParaskP7 and @AliSoftware !

While I was creating the symlinks a question came up: in `src/main/res` we have `values-in`, which is a symlink to `values-id`
image
and we have `values-iw`, which is a symlink to `values-he`
image

In `src/jetpack/res` we have the directories `values-id` and `values-he` but we **don't have** `values-in` or `values-iw`. If we do the opposite of `main/res`, considering the existing directories, we'd end up with something like:
image
image

So, is this what we're looking for? Am I breaking any scripts we might have if the symlinks are "inverted"? If so, should we rename the existing `src/jetpack/res/values-id` to `src/jetpack/res/values-in` and `src/jetpack/res/values-he` to `src/jetpack/res/values-iw` and then proceed with the symlink creation, so it looks exactly like `src/main/res`?

**Comment by AliSoftware** [comment]
@RenanLukas good question. So, iinm:

In theory I think would work either way, because what the tooling does is downloading the strings file from GlotPress for the list of locales defined [here](https://github.com/wordpress-mobile/WordPress-Android/blob/ad4e0b38e87c86177ecae94e54e8c7621499fcaa/fastlane/lanes/localization.rb#L10-L64), then writing them in `src/jetpack/res/values-{locale}/strings.xml` **file**.
 - So given that [the locale codes we use in those constants are `he` and `id`](https://github.com/wordpress-mobile/WordPress-Android/blob/ad4e0b38e87c86177ecae94e54e8c7621499fcaa/fastlane/lanes/localization.rb#L17-L18), that means for Indonesian it will write to the file in `src/jetpack/res/values-id/strings.xml` file path.
 - If `values-id` is a real folder and `values-id/strings.xml` the real file, then it will go as you expect
 - If `values-id` is the symlink pointing to `values-in`… then `values-id/strings.xml` will be the same inode and point to the same file  as `values-in/strings.xml` too… so that will work too!

In practice though, I feel it would be nice to have consistency, and always use either `values-{legacy-code} -> values-{new-code}` or `values-{new-code} -> values-{legacy-code}` in all situations, just to avoid adding confusion and inconsistencies in the mix.
I don't think one is fundamentally better than the other (I can see arguments for both) and given both should work without breaking the tooling that shouldn't matter, but better use the same convention for all cases.

---

Finally, in order to test that the tooling still works as expected, here's how you can test this:

 - Remove all the translation files in `src/main/values-*/strings.xml` and `src/jetpack/values-*/strings.xml` (e.g. `find WordPress/src/{main,jetpack}/res/values-* -name strings.xml -delete`. Or alternatively, replace all those files with empty `xml` files or something like that.
 - Run `bundle exec fastlane download_translations` to have the tooling re-download those translations from GlotPress into those `values-*` files
 - Check that the ones in `values-iw`, `values-he`, `values-id` and `values-in` are as expected, including the symlink ones.

**Comment by AliSoftware** [comment]
> If we do the opposite of main/res, considering the existing directories

@RenanLukas  Actually after checking the content of the folders, and I'm not sure I see the inconsistency you're alluding to 🤔 :

In `src/main`:
```bash
$ ls -l WordPress/src/main/res/ | grep -E 'id|in|iw|he'
drwxr-xr-x    2 olivierhalligon  staff     64 May 19 10:15 values-he
drwxr-xr-x    2 olivierhalligon  staff     64 May 19 10:15 values-id
lrwxr-xr-x    1 olivierhalligon  staff      9 Jun 16  2022 values-in -> values-id
lrwxr-xr-x    1 olivierhalligon  staff      9 Jun 16  2022 values-iw -> values-he
```
That means that in `main`, we have `values-he` and `values-id` being the "real" folders (and `values-in` and `values-iw` being symlinks to those) 

In `src/jetpack`:
```bash
$ ls -l WordPress/src/jetpack/res/ | grep -E 'id|in|iw|he'
drwxr-xr-x   2 olivierhalligon  staff   64 May 19 10:15 values-he
drwxr-xr-x   2 olivierhalligon  staff   64 May 19 10:15 values-id

```
That means that the existing directories are already the `values-he` and `values-id` real folders. So those are the same "real" folders than the ones used in `main`.

All that is missing is adding `values-in` and `values-iw` as symlink pointing to those, and we'd have the exact same setup as in `main`:
```bash
$ cd WordPress/src/jetpack/res
$ ln -s values-id values-in
$ ln -s values-he values-iw
```

So not sure what made you say above that we'd have an opposite setup for jetpack in the first place? ☕

**Comment by ParaskP7** [comment]
> In practice though, I feel it would be nice to have consistency, and always use either values-{legacy-code} -> values-{new-code} or values-{new-code} -> values-{legacy-code} in all situations, just to avoid adding confusion and inconsistencies in the mix.

I agree! 💯 

> I don't think one is fundamentally better than the other (I can see arguments for both) and given both should work without breaking the tooling that shouldn't matter, but better use the same convention for all cases.

Better use the same convention, yes! 💯

**Comment by RenanLukas** [comment]
Thanks @AliSoftware ! PR was updated.

I've tried running `bundle exec fastlane download_translations` and got the following error:

Screen Shot 2023-05-19 at 12 57 33 PM

I've tried updating the hardcoded list [here](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/fastlane/lanes/localization.rb#L68) but something was off, because even after updating it I was still getting the same error.

I've added a log right before [the line that compares expected locales with resource configs](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/fastlane/lanes/localization.rb#L389), and it seems like it didn't consider my local test (I've added `in` to the `expected_locales`, for example)
```
expected_locales: ["ar", "de", "es", "fr", "he", "id", "it", "ja", "ko", "nl", "pt-rBR", "ru", "sv", "tr", "zh-rCN", "zh-rTW"]
resource_configs: ["ar", "de", "es", "fr", "he", "id", "in", "it", "iw", "ja", "ko", "nl", "pt-rBR", "ru", "sv", "tr", "zh-rCN", "zh-rTW"]
```

**Comment by AliSoftware** [comment]
Ah! Good catch!

Adding `in` inside the `ALL_LOCALES.select { … }` won't get you the `in` locale in the `JP_APP_LOCALES` constant, because `ALL_LOCALES` doesn't contain an entry for `in` in the first place. So that's why when we're filtering that `APP_LOCALES` giant list with the `.select` filtering code, you still won't get `in` out of that.

_This part of the code and how that list of locale is defined is admittedly a bit ugly and not nice to read (I've been wanting to make it way more readable and concise for a while, even had a draft of some nice changes I worked on during a past hack week, but never found the time to clean it up)_

In practice we **don't** want to add the entry for `in` in the list of `JP_APP_LOCALES` because that constant is used to know what locales to download from GlotPress (and where to write the corresponding `values-*/strings.xml` files for each locale in the repo). But if we added `in` (and `iw`) to the `ALL_LOCALES` in the hope to fix the `resourceConfigurations` check… then that means the tooling would download the Indonesian locale from GlotPress twice, once to write into `values-id/strings.xml` and a second time doing the exact same export from the exact same GlotPress locale but to write it into `values-in/strings.xml`. Which would be a waste of API calls, and useless given we made one a symlink of the other exactly to avoid having to do that 😉 

---

So the solution in this case would be to fix the code that checks the consistency of `resourceConfigurations` to account for those symlink'd locales. One way to do that would be to modify [this line](https://github.com/wordpress-mobile/WordPress-Android/blob/bb5ac781c39d8d16ec0a46f289b5e21ce2025926/fastlane/lanes/localization.rb#L388) to instead something that that adds the old codes if the array contains the new codes:
```ruby
expected_locales = locales_list.map { |l| l[:android] }
# Support for legacy locale codes
expected_locales << 'in' if expected_locales.include?('id')
expected_locales << 'iw' if expected_locales.include?('he')
expected_locales.sort!
```

An even better / more flexible solution would be to loop thru all symlinks in the resource dirs and add those:
```ruby
Dir['WordPress/main/res/values-*'].each do |dir|
  next unless File.symlink?(dir)
  new_code = File.readlink(dir).delete_prefix('values-')
  legacy_code = dir.delete_prefix('values-')
  expected_locales << legacy_code if expected_locales.include?(new_code)
end
```
We'd have to adapt this from `WordPress/main/res/values-*` to `WordPress/jetpack/res/values-*` depending on the `app_flavor` we're checking tho. So not sure it's worth the hassle compared to hardcoding the two cases for Indonesian and Hebrew directly 🙃

**Comment by ParaskP7** [comment]
> We'd have to adapt this from WordPress/main/res/values-* to WordPress/jetpack/res/values-* depending on the app_flavor we're checking tho. So not sure it's worth the hassle compared to hardcoding the two cases for Indonesian and Hebrew directly 🙃

I agree! 💯 

Especially given the fact that we might drop those legacy codes in the (near?) future, I think it is better to hardcode the two cases for now.

**Comment by thomashorta** [comment]
Thanks, @AliSoftware and @ParaskP7 for the review and suggestions. 

I am taking care of the PR while @RenanLukas is busy with other PRs. I just implemented the "hardcoded version" of the code suggested by @AliSoftware [here](https://github.com/wordpress-mobile/WordPress-Android/pull/18455#issuecomment-1554891497). I believe it makes sense for this scenario as it's, hopefully, something temporary.

```ruby
expected_locales = locales_list.map { |l| l[:android] }
# Support for legacy locale codes
expected_locales << 'in' if expected_locales.include?('id')
expected_locales << 'iw' if expected_locales.include?('he')
expected_locales.sort!
```

The lane succeeds with these changes, thanks again!

**Review by ParaskP7 — COMMENTED** [review]
👋 @thomashorta and thanks for taking care of this PR on behalf of @RenanLukas ! 🥇 

> I just implemented the "hardcoded version" of the code suggested by @AliSoftware https://github.com/wordpress-mobile/WordPress-Android/pull/18455#issuecomment-1554891497.

👍 

> I believe it makes sense for this scenario as it's, hopefully, something temporary.

I agree! 💯 

> The lane succeeds with these changes, thanks again!

Awesome! 🎉 

Btw, I am seeing this 0c112c62154621693098073ee4de8c7520980223 and acb1ac215c8846d4176b8145ef2374341c2d9163 commits included in this PR. I understand that those got added automatically by running fastlane. That was done mainly for testing purposes, just to verify that the lane is working as expected, right? 🤔

Having said that, I am not sure if we want to include those commits or completely drop them, just to avoid any confusion/conflicts later on and during normal release management, not sure...

-----

From my side we are good to go with this PR, but I'll be letting @AliSoftware have the final say (✅) before merging this to `trunk`, just in case.

File: WordPress/build.gradle
Comment (AliSoftware) [inline WordPress/build.gradle]: Maybe also add a note as code-comment about legacy locales?

```suggestion
            // Note 1: this only affects included _locales_; variants for `values-night`, `values-land` and other configs are _still_ preserved in Jetpack and not filtered by this.
            // Note 2: `in` is the legacy locale code for `id` and `iw` is the legacy locale code for `he`. 
            // We need to include those to make sure users with older Android versions and still using those legacy codes still have those translations show up at runtime.
            resourceConfigurations = ["ar", "de", "es", "fr", "he", "id", "in", "it", "iw", "ja", "ko", "nl", "pt-rBR", "ru", "sv", "tr", "zh-rCN", "zh-rTW"]
```

**Review by AliSoftware — COMMENTED** [review]
I agree with @ParaskP7 , better drop the testing commits to avoid this PR including any changes in the `strings.xml` files.
 - Not only because it's not the goal of this PR to update the translations themselves to their latest values
 - But also because we don't want it to risk creating conflict with the next time we pull the latest translations into the release branch at end of sprint as part of the regular release management process and then have it risk conflict with the entries in `develop` that this PR would have included.

I've also left a suggestion to comment about legacy locale codes on `resourcesConfigurations` line in `build.gradle` too.

Apart from that, once those commits are removed from the PR we should be good to merge!

File: WordPress/build.gradle
Comment (thomashorta) [reply WordPress/build.gradle]: That makes sense! I added the comments and also added the legacy codes in the list in a separate line, so it's clear we're adding them as additional requirements (with the comment making that 200% clear).

**Comment by thomashorta** [comment]
> Btw, I am seeing this [0c112c6](https://github.com/wordpress-mobile/WordPress-Android/commit/0c112c62154621693098073ee4de8c7520980223) and [acb1ac2](https://github.com/wordpress-mobile/WordPress-Android/commit/acb1ac215c8846d4176b8145ef2374341c2d9163) commits included in this PR. I understand that those got added automatically by running fastlane. That was done mainly for testing purposes, just to verify that the lane is working as expected, right? 🤔

Thanks for catching that @ParaskP7 and @AliSoftware! I didn't notice the automatically created commits, my bad 🤦. I removed them from the branch.

I also added @AliSoftware's suggestion regarding comments and made it clear that we are adding these legacy codes because of compatibility reasons.

**Review by AliSoftware — APPROVED** [review]

**Review by ParaskP7 — APPROVED** [review]

**Pull request merged**
